opm/simulators/flow/OutputBlackoilModule.hpp

```diff
@@ -1775,8 +1564,340 @@ class OutputBlackOilModule : public GenericOutputBlackoilModule<GetPropType<Type
+                             [&problem = this->simulator_.problem()](const Context& ectx)
+                             {
+                                 const auto& materialParams = problem.materialLawParams(ectx.elemCtx,
+                                                                                          ectx.dofIdx,
+                                                                                          /* timeIdx = */ 0);
+                                 return getValue(MaterialLaw::template
+                                                     relpermOilInOilWaterSystem<Evaluation>(materialParams,
+                                                                                            ectx.fs));
+                            }
+                 }
+            },
+            Entry{ScalarEntry{"BWPC",
+                              [](const Context& ectx)
+                              {
+                                  return getValue(ectx.fs.pressure(oilPhaseIdx)) -
+                                         getValue(ectx.fs.pressure(waterPhaseIdx));
+                              }
+                  }
+            },
+            Entry{ScalarEntry{"BGPC",
+                              [](const Context& ectx)
+                              {
+                                  return getValue(ectx.fs.pressure(gasPhaseIdx)) -
+                                         getValue(ectx.fs.pressure(oilPhaseIdx));
+                              }
+                  }
+            },
+            Entry{ScalarEntry{"BWPR",
+                              [](const Context& ectx)
+                              {
+                                  return getValue(ectx.fs.pressure(waterPhaseIdx));
+                              }
+                  }
+            },
+            Entry{ScalarEntry{"BGPR",
+                              [](const Context& ectx)
+                              {
+                                  return getValue(ectx.fs.pressure(gasPhaseIdx));
+                              }
+                  }
+            },
+            Entry{PhaseEntry{std::array{
+                                std::array{"BVWAT"sv, "BVOIL"sv, "BVGAS"sv},
+                                std::array{"BWVIS"sv, "BOVIS"sv, "BGVIS"sv}
+                             },
+                             [](const unsigned phaseIdx, const Context& ectx)
+                             {
+                                 return getValue(ectx.fs.viscosity(phaseIdx));
+                             }
+                  }
+            },
+            Entry{PhaseEntry{std::array{
+                                std::array{"BWDEN"sv, "BODEN"sv, "BGDEN"sv},
+                                std::array{"BDENW"sv, "BDENO"sv, "BDENG"sv}
+                             },
+                             [](const unsigned phaseIdx, const Context& ectx)
+                             {
+                                 return getValue(ectx.fs.density(phaseIdx));
```

Comment:
This, `fs.density(phaseIdx)` and other statements like it, only work by serendipitous circumstance, and probably only for black-oil type simulation runs since the phase index value generally depends on the `FluidSystem`.  On the other hand, `fs.density(gasPhaseIdx)` is safe, so one option might be to have a layer of indirection that translates the numbers 0,1,2 to `waterPhaseIdx`, `oilPhaseIdx`, and `gasPhaseIdx`, respectively.

Reply:
yeah, you are right even, the canonical phase indices are the same for all fluid systems in use so i took a shortcut. added a translation.